### ISSUE TITLE
TimeHelper: Remove all deprecated methods by updating their callers to call replacement methods instead

### ISSUE THREAD

**Issue by darrenwee** [TEAMMATES/teammates#8645]
All `@Deprecated` methods should be removed by the end. This can be done through small PRs made directly against `master`

From now on, PRs against master that use `TimeHelper` in new code should not call the old methods anymore; they should instead call the new methods

**Comment by whipermr5** [comment]
This was what I was working on for your reference. I'll leave this issue to you.

```patch
From 255b69426097bae8199a5eb1dd825cf9cf83e8ae Mon Sep 17 00:00:00 2001
From: John Yong <whipermr5@gmail.com>
Date: Thu, 15 Mar 2018 05:18:37 +0800
Subject: [PATCH] Remove usages of Date API from HtmlHelper

Also fix test failure during period of the day when date in UTC differs from date in local time zone
---
 .../java/teammates/test/driver/HtmlHelper.java     | 31 ++++++++++------------
 1 file changed, 14 insertions(+), 17 deletions(-)

diff --git a/src/test/java/teammates/test/driver/HtmlHelper.java b/src/test/java/teammates/test/driver/HtmlHelper.java
index b81f22e..5a41e20 100644
--- a/src/test/java/teammates/test/driver/HtmlHelper.java
+++ b/src/test/java/teammates/test/driver/HtmlHelper.java
@@ -2,8 +2,9 @@ package teammates.test.driver;
 
 import static org.testng.AssertJUnit.assertEquals;
 
-import java.text.SimpleDateFormat;
-import java.util.Date;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
 import java.util.TimeZone;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -411,16 +412,12 @@ public final class HtmlHelper {
      * These values are identified using their known, unique formats.
      */
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
-        Date now = new Date();
-        SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        Instant now = Instant.now();
         // get session's time zone from content.
         // this method is not applicable for pages with multiple time zones like InstructorSearchPage
-        sdf.setTimeZone(getTimeZone(content));
-        String dateTimeNow = sdf.format(now);
-        SimpleDateFormat sdfForIso8601 = new SimpleDateFormat("yyyy-MM-dd'T'");
-        String dateTimeNowInIso8601 = sdfForIso8601.format(now);
-        SimpleDateFormat sdfForCoursesPage = new SimpleDateFormat("d MMM yyyy");
-        String dateTimeNowInCoursesPageFormat = sdfForCoursesPage.format(now);
+        String dateTimeNow = TimeHelper.formatTime12H(LocalDateTime.now(getTimeZone(content))).substring(0, 18);
+        String dateTimeNowInIso8601 = TimeHelper.formatInstantToIso8601Utc(now).substring(0, 11);
+        String dateTimeNowInCoursesPageFormat = TimeHelper.formatDateTimeForInstructorCoursesPage(now, ZoneId.of("UTC"));
         return content // dev server admin absolute URLs (${teammates.url}/_ah/...)
                       .replace("\"" + TestProperties.TEAMMATES_URL + "/_ah", "\"/_ah")
                       // logout URL generated by Google
@@ -510,7 +507,7 @@ public final class HtmlHelper {
      * This method is only used for testing.
      */
     public static String injectContextDependentValuesForTest(String content) {
-        Date now = new Date();
+        Instant now = Instant.now();
         return content.replace("<!-- test.url -->", TestProperties.TEAMMATES_URL)
                       .replace("<!-- studentmotd.url -->", Config.STUDENT_MOTD_URL)
                       .replace("<!-- support.email -->", Config.SUPPORT_EMAIL)
@@ -527,14 +524,14 @@ public final class HtmlHelper {
                       .replace("<!-- test.admin -->", TestProperties.TEST_ADMIN_ACCOUNT)
                       .replace("<!-- test.admin.truncated -->",
                                StringHelper.truncateLongId(TestProperties.TEST_ADMIN_ACCOUNT))
-                      .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.sessions -->", TimeHelper.formatDateTimeForSessions(now, 0))
-                      .replace("<!-- now.datetime.iso8601utc -->", TimeHelper.formatDateToIso8601Utc(now))
+                      .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(LocalDateTime.now(ZoneId.of("UTC"))))
+                      .replace("<!-- now.datetime.sessions -->", TimeHelper.formatDateTimeForSessions(now, ZoneId.of("UTC")))
+                      .replace("<!-- now.datetime.iso8601utc -->", TimeHelper.formatInstantToIso8601Utc(now))
                       .replace("<!-- now.datetime.courses -->",
-                              TimeHelper.formatDateTimeForInstructorCoursesPage(now, "UTC"));
+                              TimeHelper.formatDateTimeForInstructorCoursesPage(now, ZoneId.of("UTC")));
     }
 
-    private static TimeZone getTimeZone(String content) {
+    private static ZoneId getTimeZone(String content) {
         // searches for first String of pattern "UTC+xxxx" in the content.
         Pattern pattern = Pattern.compile(REGEX_TIMEZONE_OFFSET);
         Matcher matcher = pattern.matcher(content);
@@ -543,7 +540,7 @@ public final class HtmlHelper {
         if (matcher.find()) {
             timeZoneOffset = matcher.group(1);
         }
-        return TimeZone.getTimeZone("GMT" + timeZoneOffset);
+        return TimeZone.getTimeZone("GMT" + timeZoneOffset).toZoneId();
     }
 
 }
-- 
2.10.1

From 17e8aa91453a5258732c082ca2104abfdb76b50a Mon Sep 17 00:00:00 2001
From: John Yong <whipermr5@gmail.com>
Date: Thu, 15 Mar 2018 05:19:09 +0800
Subject: [PATCH] Remove usages of Date API from TimeHelperTest

---
 .../teammates/test/cases/util/TimeHelperTest.java  | 60 ++++++++++------------
 1 file changed, 27 insertions(+), 33 deletions(-)

diff --git a/src/test/java/teammates/test/cases/util/TimeHelperTest.java b/src/test/java/teammates/test/cases/util/TimeHelperTest.java
index 6cc9d40..ad84a1b 100644
--- a/src/test/java/teammates/test/cases/util/TimeHelperTest.java
+++ b/src/test/java/teammates/test/cases/util/TimeHelperTest.java
@@ -1,9 +1,11 @@
 package teammates.test.cases.util;
 
+import java.time.Instant;
+import java.time.LocalDate;
 import java.time.LocalDateTime;
-import java.util.Calendar;
-import java.util.Date;
-import java.util.TimeZone;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
 
 import org.testng.annotations.Test;
 
@@ -54,44 +56,36 @@ public class TimeHelperTest extends BaseTestCase {
 
     @Test
     public void testEndOfYearDates() {
-        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
-        cal.clear();
-        cal.set(2015, 11, 30, 12, 0, 0);
-        Date date = cal.getTime();
+        LocalDateTime date = LocalDateTime.of(LocalDate.of(2015, 12, 30), LocalTime.NOON);
         assertEquals("30/12/2015", TimeHelper.formatDate(date));
         assertEquals("Wed, 30 Dec, 2015", TimeHelper.formatDateForSessionsForm(date));
         assertEquals("Wed, 30 Dec 2015, 12:00 NOON", TimeHelper.formatTime12H(date));
-        assertEquals("Wed, 30 Dec 2015, 12:00 NOON UTC+0000", TimeHelper.formatDateTimeForSessions(date, 0));
+        assertEquals("Wed, 30 Dec 2015, 12:00 NOON UTC+0000",
+                TimeHelper.formatDateTimeForSessions(date.toInstant(ZoneOffset.UTC), ZoneId.of("UTC")));
         assertEquals("30 Dec 12:00 NOON", TimeHelper.formatDateTimeForInstructorHomePage(date));
     }
 
     @Test
     public void testFormatDateTimeForSessions() {
-        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
-        cal.clear();
-        cal.set(2015, 10, 30, 12, 0, 0);
-        Date date = cal.getTime();
-        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC+0000", TimeHelper.formatDateTimeForSessions(date, 0));
-
-        cal.clear();
-        cal.set(2015, 10, 30, 4, 0, 0);
-        date = cal.getTime();
-        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC+0800", TimeHelper.formatDateTimeForSessions(date, 8));
-
-        cal.clear();
-        cal.set(2015, 10, 30, 4, 0, 0);
-        date = cal.getTime();
-        assertEquals("Mon, 30 Nov 2015, 04:00 PM UTC+1200", TimeHelper.formatDateTimeForSessions(date, 12));
-
-        cal.clear();
-        cal.set(2015, 10, 30, 16, 0, 0);
-        date = cal.getTime();
-        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC-0400", TimeHelper.formatDateTimeForSessions(date, -4));
-
-        cal.clear();
-        cal.set(2015, 10, 30, 16, 0, 0);
-        date = cal.getTime();
-        assertEquals("Mon, 30 Nov 2015, 11:45 AM UTC-0415", TimeHelper.formatDateTimeForSessions(date, -4.25));
+        LocalDate localDate = LocalDate.of(2015, 11, 30);
+        Instant instant = LocalDateTime.of(localDate, LocalTime.of(12, 0)).toInstant(ZoneOffset.UTC);
+
+        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC+0000",
+                TimeHelper.formatDateTimeForSessions(instant, ZoneId.of("UTC")));
+
+        instant = LocalDateTime.of(localDate, LocalTime.of(4, 0)).toInstant(ZoneOffset.UTC);
+
+        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC+0800",
+                TimeHelper.formatDateTimeForSessions(instant, ZoneId.of("UTC+08:00")));
+        assertEquals("Mon, 30 Nov 2015, 04:00 PM UTC+1200",
+                TimeHelper.formatDateTimeForSessions(instant, ZoneId.of("UTC+12:00")));
+
+        instant = LocalDateTime.of(localDate, LocalTime.of(16, 0)).toInstant(ZoneOffset.UTC);
+
+        assertEquals("Mon, 30 Nov 2015, 12:00 NOON UTC-0400",
+                TimeHelper.formatDateTimeForSessions(instant, ZoneId.of("UTC-04:00")));
+        assertEquals("Mon, 30 Nov 2015, 11:45 AM UTC-0415",
+                TimeHelper.formatDateTimeForSessions(instant, ZoneId.of("UTC-04:15")));
     }
 
 }
-- 
2.10.1

From 82510586ba7866c204ad9c4946602e451c7fe672 Mon Sep 17 00:00:00 2001
From: John Yong <whipermr5@gmail.com>
Date: Thu, 15 Mar 2018 05:19:25 +0800
Subject: [PATCH] Remove now unused TimeHelper deprecated methods

---
 .../java/teammates/common/util/TimeHelper.java     | 41 ----------------------
 1 file changed, 41 deletions(-)

diff --git a/src/main/java/teammates/common/util/TimeHelper.java b/src/main/java/teammates/common/util/TimeHelper.java
index 131786a..f5fd333 100644
--- a/src/main/java/teammates/common/util/TimeHelper.java
+++ b/src/main/java/teammates/common/util/TimeHelper.java
@@ -319,12 +319,6 @@ public final class TimeHelper {
         return zonedDateTime.format(formatter);
     }
 
-    @Deprecated
-    public static String formatDateTimeForSessions(Date dateInUtc, double sessionTimeZone) {
-        return formatDateTimeForSessions(
-                convertDateToInstant(dateInUtc), convertToZoneId(sessionTimeZone));
-    }
-
     public static String formatDateTimeForSessions(Instant instant, ZoneId sessionTimeZone) {
         return formatInstant(instant, sessionTimeZone, "EEE, dd MMM yyyy, hh:mm a 'UTC'Z");
     }
@@ -332,14 +326,6 @@ public final class TimeHelper {
     /**
      * Formats a date in the format d MMM h:mm a. Example: 5 May 11:59 PM
      */
-    @Deprecated
-    public static String formatDateTimeForInstructorHomePage(Date date) {
-        return formatDateTimeForInstructorHomePage(convertDateToLocalDateTime(date));
-    }
-
-    /**
-     * Formats a date in the format d MMM h:mm a. Example: 5 May 11:59 PM
-     */
     public static String formatDateTimeForInstructorHomePage(LocalDateTime localDateTime) {
         return formatLocalDateTime(localDateTime, "d MMM h:mm a");
     }
@@ -347,27 +333,11 @@ public final class TimeHelper {
     /**
      * Formats a date in the format d MMM yyyy. Example: 5 May 2017
      */
-    @Deprecated
-    public static String formatDateTimeForInstructorCoursesPage(Date date, String timeZoneId) {
-        return formatDateTimeForInstructorCoursesPage(convertDateToInstant(date), ZoneId.of(timeZoneId));
-    }
-
-    /**
-     * Formats a date in the format d MMM yyyy. Example: 5 May 2017
-     */
     public static String formatDateTimeForInstructorCoursesPage(Instant instant, ZoneId timeZoneId) {
         return formatInstant(instant, timeZoneId, "d MMM yyyy");
     }
 
     /**
-     * Formats {@code dateInUtc} according to the ISO8601 format.
-     */
-    @Deprecated
-    public static String formatDateToIso8601Utc(Date dateInUtc) {
-        return formatInstantToIso8601Utc(convertDateToInstant(dateInUtc));
-    }
-
-    /**
      * Formats {@code instant} according to the ISO8601 format.
      */
     public static String formatInstantToIso8601Utc(Instant instant) {
@@ -379,17 +349,6 @@ public final class TimeHelper {
     }
 
     /**
-     * Converts the date string to a Date object.
-     *
-     * @param dateInStringFormat should be in the format {@link SystemParams#DEFAULT_DATE_TIME_FORMAT}
-     * @deprecated Use {@link TimeHelper#parseInstant(String)} instead
-     */
-    @Deprecated
-    public static Date convertToDate(String dateInStringFormat) {
-        return convertInstantToDate(parseInstant(dateInStringFormat));
-    }
-
-    /**
      * Converts the datetime string to an Instant object.
      *
      * @param dateTimeString should be in the format {@link SystemParams#DEFAULT_DATE_TIME_FORMAT}
-- 
2.10.1

```

**Comment by tran-tien-dat** [comment]
Oops. Seems like my new issue #8731 and PR #8732 is a duplicate of this 😝 . However, what I have done is to search for every occurences of `import java.util.Date` etc. and change them to use `java.time`. The next thing would be to migrate the `TimeHelper` class, which @whipermr5 has done here.

**Comment by darrenwee** [comment]
Resolved by #8732
